src/main.rs

```diff
@@ -0,0 +1,127 @@
+// Copyright 2018 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not
+// use this file except in compliance with the License.  You may obtain a copy
+// of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+// License for the specific language governing permissions and limitations
+// under the License.
+
+extern crate env_logger;
+extern crate getopts;
+extern crate sandboxfs;
+
+use getopts::Options;
+use std::env;
+use std::ffi::OsStr;
+use std::io;
+use std::path::Path;
+use std::process;
+use std::result::Result;
+
+/// Error-like type to encapsulate various different error conditions in the
+/// main program.
+#[derive(Debug)]
+enum MainError {
+    /// Execution failed due to a generic runtime error.
+    Runtime(String),
+    /// Execution failed due to a user-triggered error.
+    Usage(String),
+}
+
+impl From<io::Error> for MainError {
+    fn from(err: io::Error) -> Self {
+        MainError::Runtime(format!("{}", err))
+    }
+}
+
+impl From<getopts::Fail> for MainError {
+    fn from(err: getopts::Fail) -> Self {
+        MainError::Usage(format!("{}", err))
+    }
+}
+
+/// Obtains the program name from the execution's first argument, or returns a
+/// default if the program name cannot be determined for whatever reason.
+fn program_name(args: &[String], default: &'static str) -> String {
+    args.get(0)
```

Comment:
Long chains like this are hard to read. Instead, break out of the monad early and use regular code.

Reply:
Good. Somehow I thought this was common style in Rust as I've seen it in many places... but glad to see I'm not the only one thinking it's hard to read.

I've replaced this with a match chain... which maybe is also hard?